sync-todo/v2/client/cpp/managers/database_manager.cpp

```diff
@@ -0,0 +1,151 @@
+#include "database_manager.hpp"
+
+DatabaseManager::DatabaseManager(AppState *appState): _appState(appState) {
+  auto user = _appState->app->get_current_user();
+
+  /** Sync configuration and sync subscription management. */
+  allItemSubscriptionName = "all_items";
+  myItemSubscriptionName = "my_items";
+  userId = user->identifier();
+
+  auto config = user->flexible_sync_configuration();
+
+  /** Handle database sync errors. If there is an error, add the message to the `appState`
+   * and change the screen that is displaying to the error modal. */
+  config.sync_config().set_error_handler([=](const realm::sync_session &session,
+                                             const realm::internal::bridge::sync_error &error) {
+
+    auto errorText = SS("A sync error occurred. Message: " << error.message() << std::endl);
+    _appState->errorManager->setError(errorText);
+  });
+
+  /** Initialize the database, and add a subscription to all items. This enables the app to read
+   *  all items in the data source linked to the App Services App. App Services Rules for the Item collection
+   *  mean that while the user can read all items, they can only write to their own items. */
+  auto database = realm::db(std::move(config));
+  database.subscriptions().update([&](realm::mutable_sync_subscription_set& subs) {
+    /** By default, we show all items.*/
+    if (!subs.find(allItemSubscriptionName)) {
+      subs.add<realm::Item>(allItemSubscriptionName);
+    }
+  }).get();
+
+  /** Wait for downloads after loading the app, and refresh the database.*/
+  database.get_sync_session()->wait_for_download_completion().get();
+  database.refresh();
+
+  /** Add a pointer to the database as a class member, so we can access the database later when making changes. */
+  databasePtr = std::make_unique<realm::db>(database);
+}
+
+/** Add a new item to the task list. */
+void DatabaseManager::addNew() {
+  auto item = realm::Item {
+    .isComplete = _appState->databaseState->newTaskIsComplete,
```

Comment:
That's fair! I need the state object for FTXUI's input fields, to hold a persistent reference to those fields. But I can just pass the values into this function. I'm trying to avoid making the Item in the view code and instead letting the function handle it, to keep the DB logic encapsulated in this DBManager class.